File: src/main/java/net/nyavro/spring/social/signinmvc/controller/ResourceController.java
package net.nyavro.spring.social.signinmvc.controller;

import org.h2.util.IOUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;

@Controller
@RequestMapping(value = "/resource", produces = MediaType.APPLICATION_OCTET_STREAM_VALUE)
public class ResourceController {

    private static final Logger log = LoggerFactory.getLogger(ResourceController.class);

    @RequestMapping(value = "/avatar/{id}", method = RequestMethod.GET)
    public void avatar(final @PathVariable(value = "id") String id, HttpServletRequest request, HttpServletResponse response) {
        log.debug("Fetching avatar of user {}", id);
        try {
            response.setContentType("image/png");
            //Placeholder for now:
            IOUtils.copy(
                new URL(host(request) + "/img/account_100x100.png").openConnection().getInputStream(),
                response.getOutputStream()
            );
        } catch (MalformedURLException e) {
            throw new RuntimeException(e);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private String host(HttpServletRequest request) {
        return request.getScheme() + "://" + request.getLocalAddr() + ":" + request.getLocalPort();
    }
}

File: src/main/scala/com/nyavro/dienstleustigen/controller/ResourceController.scala
package com.nyavro.dienstleustigen.controller

import org.springframework.http.MediaType
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation._
import java.net.{MalformedURLException, URL}
import javax.servlet.http.{HttpServletResponse, HttpServletRequest}
import com.jcabi.log.Logger
import java.io.IOException
import org.apache.commons.io.IOUtils
import org.springframework.web.multipart.MultipartFile
import org.springframework.beans.factory.annotation.Autowired
import com.nyavro.dienstleustigen.service.FileStorage

@Controller
@RequestMapping(value = Array("/resource"), produces = Array(MediaType.APPLICATION_OCTET_STREAM_VALUE))
class ResourceController {

  @Autowired
  var storage:FileStorage = _

  @RequestMapping(value = Array("/avatar/{id}"), method = Array(RequestMethod.GET))
  def avatar(@PathVariable(value = "id") id: String, request: HttpServletRequest, response: HttpServletResponse) = {
    Logger.debug(this, "Fetching avatar of user {}", id)
    response.setContentType("image/png")
    IOUtils.copy(new URL(host(request) + "/img/account_100x100.png").openConnection.getInputStream, response.getOutputStream)
  }
  
  @RequestMapping(value = Array("/upload"), method = Array(RequestMethod.POST))
  @ResponseBody def uploadFile(@RequestParam("file") file: MultipartFile): Map[String, String] = {
    try {
      Map("status" -> "OK", "id" -> storage.upload(file.getInputStream))
    }
    catch {
      case e: IOException => Map("status" -> "error", "message" -> ("Unable to store file: " + e.getMessage))
      case e: Any => throw new RuntimeException(e)
    }
  }

  @RequestMapping(value = Array("/download"), method = Array(RequestMethod.GET))
  def downloadFile(@RequestParam(value = "id") id: String, request: HttpServletRequest, response: HttpServletResponse) = {
    Logger.debug(this, "Fetching file {}", id)
    storage.download(id, response.getOutputStream)
  }

  private def host(request: HttpServletRequest): String = request.getScheme + "://" + request.getLocalAddr + ":" + request.getLocalPort
}
